File: Marrubium.Web/Services/ProductService.cs
using Marrubium.Web.Models;
using Marrubium.Web.Services.IServices;

namespace Marrubium.Web.Services
{
    public class ProductService : BaseService, IProductService
    {
        private readonly IHttpClientFactory _clientFactory;

        public ProductService(IHttpClientFactory clientFactory) : base(clientFactory)
        {
            _clientFactory = clientFactory;
        }

        async Task<T> IProductService.CreateProductAsync<T>(ProductDto productDto)
        {
            return await this.SendAsync<T>(new ApiRequest
            {
                ApiType = SD.ApiType.POST,
                Data = productDto,
                Url = SD.ProductAPIBase + "/api/products",
                AccessToken = ""
            });
        }

        async Task<T> IProductService.DeleteProductAsync<T>(int id)
        {
            return await this.SendAsync<T>(new ApiRequest
            {
                ApiType = SD.ApiType.DELETE,
                Url = SD.ProductAPIBase + $"/api/products/{id}",
                AccessToken = ""
            });
        }

        async Task<T> IProductService.GetProductByIdAsync<T>(int id)
        {
            return await this.SendAsync<T>(new ApiRequest
            {
                ApiType = SD.ApiType.GET,
                Url = SD.ProductAPIBase + $"/api/products/{id}",
                AccessToken = ""
            });
        }

        async Task<T> IProductService.GetAllProductsAsync<T>()
        {
            return await this.SendAsync<T>(new ApiRequest
            {
                ApiType = SD.ApiType.GET,
                Url = SD.ProductAPIBase + "/api/products",
                AccessToken = ""
            });
        }

        async Task<T> IProductService.UpdateProductAsync<T>(ProductDto productDto)
        {
            return await this.SendAsync<T>(new ApiRequest
            {
                ApiType = SD.ApiType.PUT,
                Data = productDto,
                Url = SD.ProductAPIBase + "/api/products",
                AccessToken = ""
            });
        }
    }
}

File: Marrubium.Web/TypeScriptData/Services/ProductService.ts
import {IProductService} from "./IServices/IProductService.js";
import {BaseService} from "./BaseService.js";
import {ApiMethod} from "../Models/ApiMethod.js";
import {ApiRequest} from "../Models/ApiRequest.js";
import {ProductDto} from "../Models/ProductDto.js";
import {CategoryListsModel} from "../Models/CategoryListsModel.js";


export class ProductService extends BaseService implements IProductService {
    
    constructor(baseUrl: string)  {
        super(baseUrl)
    }

    getProductByIdAsync(productId: number): Promise<ProductDto> {
        return this.sendAsync(new ApiRequest(
            ApiMethod.DELETE,
            `api/products/productId=${productId}`
        ))
            .then(x => Object.assign(new ProductDto(), x));
    }

    getProductsAsync(): Promise<Array<ProductDto>> {
        return this.sendAsync(new ApiRequest(
            ApiMethod.GET,
            "api/products"
        ))
            .then(x => Object.assign(new Array<ProductDto>(), x));
    }

    getProductCategoriesAsync(): Promise<CategoryListsModel> {
        return this.sendAsync(new ApiRequest(
            ApiMethod.GET,
            "api/products/categories"
        ))
            .then(x => Object.assign(new CategoryListsModel(), x));
    }
}